src/Database/PG/Query/Connection.hs

```diff
@@ -365,20 +369,48 @@ execPrepared
   :: PQ.Connection                -- ^ connection
   -> [Maybe (DB.ByteString, PQ.Format)]           -- ^ parameters
   -> RemoteKey                    -- ^ stmtName
+  -> ((PQ.Connection -> IO (Maybe PQ.Result)) -> PQ.Connection -> IO (Maybe PQ.Result))
   -> PGExec ResultOk -- ^ result
-execPrepared conn args n = do
-  mRes <- lift $ PQ.execPrepared conn n args PQ.Binary
+execPrepared conn args n run = do
+  mRes <- lift $ run (\c -> PQ.execPrepared c n args PQ.Binary) conn
   checkResult conn mRes
 
+data PGCancelErr = CEAllocate | CEError DT.Text
+  deriving (Show, Eq)
+
+instance Exception PGCancelErr
+
+cancelPG :: PQ.Connection -> IO (Either PGCancelErr ())
+cancelPG conn = do
+  PQ.getCancel conn >>= \case
+    Nothing -> return $ Left CEAllocate
+    Just c  -> PQ.cancel c >>= \case
+      Left err -> return $ Left $ CEError $ lenientDecodeUtf8 err
+      Right () -> return $ Right ()
+
+cancelOnTimeout :: (PQ.Connection -> IO a) -> PQ.Connection -> IO a
+cancelOnTimeout f conn = do
+  a <- async $ f conn
```

Comment:
Why is it necessary to execute in a new thread, and can this be documented in a comment?

Reply:
The new thread is the core of that hack, so to speak: That thread stays blocked in the FFI call when we receive the exception, while the `wait` call is interrupted by the async exception (not because it's rethrown from the async action, but because it's thrown at it from `timeout`). Then we can send cancel on the same connection (current implementation is buggy in that we shouldn't touch the connection from another thread, about to fix that by running `getCancel` before forking).

(That's all with the massive caveat that my mental model of the GHC runtime may be off.)

I've tried to put together a higher level overview of all the related things here: https://github.com/hasura/graphql-engine-mono/blob/4c0980e4aef3d394e047c48d3d2697718b66a681/rfcs/operation-timeout-api-limits-braindump.md.

Reply:
And yeah I'd definitely document this better, the current state is really at the point of "prove that it's possible at all to interrupt long-running queries using this method".